parcels/interaction/hash_spherical.py

```diff
@@ -0,0 +1,167 @@
+        i_depth = np.floor(depth/self.interaction_depth).astype(int)
+
+        # Get the arc length of the smaller circle around the earth.
+        circ_small = 2*np.pi*np.cos((i_lat+1)*self.inter_arc_dist)
+        n_long = np.floor(circ_small/self.inter_arc_dist).astype(int)
+        n_long[n_long < 1] = 1
+        d_long = 360/n_long
+
+        # Get the longitude part of the cell id.
+        i_long = np.floor(long/d_long).astype(int)
+
+        # Merge the 4 parts of the cell into one id.
+        point_hash = i_3d_to_hash(i_lat, i_long, i_depth, lat_sign, self._bits)
+        point_array = np.empty(values.shape[1], dtype=int)
+        point_array[active_idx] = point_hash
+        return point_array
+
+    def rebuild(self, values, active_mask=-1):
+        '''Recreate the tree with new values.
+
+        :param values: positions of the particles.
+        '''
+        super().rebuild(values, active_mask)
+        active_idx = self.active_idx
+
+        # Compute the hash values:
+        self._particle_hashes = np.empty(self._values.shape[1], dtype=int)
+        self._particle_hashes[active_idx] = self._values_to_hashes(
+            values[:, active_idx])
+
+        # Create the hashtable.
+        self._hashtable = hash_split(self._particle_hashes,
+                                     active_idx=active_idx)
+
+        # Keep track of the position of a particle index within a cell.
+        self._hash_idx = np.empty_like(self._particle_hashes, dtype=int)
+        for idx_array in self._hashtable.values():
+            self._hash_idx[idx_array] = np.arange(len(idx_array))
+
+    def _init_structure(self):
+        '''Initialize the basic tree properties without building'''
+        epsilon = 1e-12
+        R_earth = 6371000
+
+        self.inter_arc_dist = self.interaction_distance/R_earth
+        self.inter_degree_dist = 180*self.inter_arc_dist/np.pi
+        n_lines_lat = int(ceil(np.pi/self.inter_arc_dist+epsilon))
+        n_lines_long = int(ceil(2*np.pi/self.inter_arc_dist+epsilon))
+        n_lines_depth = int(ceil(
+            self.max_depth/self.interaction_depth + epsilon))
+        n_bits_lat = ceil(np.log(n_lines_lat)/np.log(2))
+        n_bits_long = ceil(np.log(n_lines_long)/np.log(2))
+        n_bits_depth = ceil(np.log(n_lines_depth)/np.log(2))
+        self._bits = np.array([n_bits_lat, n_bits_long, n_bits_depth])
+
+
+# @njit
+def i_3d_to_hash(i_lat, i_long, i_depth, lat_sign, bits):
+    '''Convert longitude and lattitude id's to hash'''
+    point_hash = lat_sign
+    point_hash = np.bitwise_or(point_hash, np.left_shift(i_lat, 1))
+    point_hash = np.bitwise_or(point_hash, np.left_shift(i_long, 1+bits[0]))
+    point_hash = np.bitwise_or(point_hash,
+                               np.left_shift(i_depth, 1+bits[0]+bits[1]))
+    return point_hash
+
+
+def geo_hash_to_neighbors(hash_id, coor, bits, inter_arc_dist):
+    '''Compute the hashes of all neighboring cells.'''
```

Comment:
"""Compute the hashes of all neighboring cells in a 3x3 block neighborhood."""

Reply:
Done